coordinator/src/main/scala/filodb.coordinator/queryplanner/LongTimeRangeSplitPlanner.scala

```diff
@@ -0,0 +1,118 @@
+          }
+          splitPlan(downsampleLp, qContext, p1 =>  downsampleClusterPlanner.materialize(p1, qContext))
+        } else {
+          // Split the query between raw and downsample planners
+          val numStepsInDownsample = (earliestRawTime - startWithOffsetMs + lookbackMs) / p.stepMs
+          val lastDownsampleInstant = p.startMs + numStepsInDownsample * p.stepMs
+          val firstInstantInRaw = lastDownsampleInstant + p.stepMs
+
+          val downsampleLp = copyWithUpdatedTimeRange(logicalPlan,
+            TimeRange(p.startMs, lastDownsampleInstant))
+          val downsampleEp = splitPlan(downsampleLp, qContext,
+            p1 => downsampleClusterPlanner.materialize(p1, qContext))
+
+          val rawLp = copyWithUpdatedTimeRange(logicalPlan, TimeRange(firstInstantInRaw, p.endMs))
+          val rawEp = splitPlan(rawLp, qContext, p1 => rawClusterPlanner.materialize(p1, qContext))
+          StitchRvsExec(qContext, stitchDispatcher, Seq(rawEp, downsampleEp))
+        }
+      // Metadata query not supported for downsample cluster
+      case _ => rawClusterPlanner.materialize(logicalPlan, qContext)
+    }
+  }
+
+  /**
+    * Split query if the time range is greater than the threshold. It clones the given LogicalPlan with the smaller
+    * time ranges, creates an execPlan using the provided planner and finally returns Stitch ExecPlan
+    * @param lPlan LogicalPlan to be split
+    * @param qContext QueryContext
+    * @param planner provided function which takes input as LogicalPlan and creates an ExecPlan
+    */
+  private def splitPlan(lPlan: LogicalPlan, qContext: QueryContext,
+                        planner: PeriodicSeriesPlan => ExecPlan): ExecPlan = {
+    val windowSplits: Buffer[(Long, Long)] = Buffer.empty
+    val lp = lPlan.asInstanceOf[PeriodicSeriesPlan]
+    if (lp.isSplittable && lp.endMs - lp.startMs > splitThresholdMs) {
+      logger.info(s"Splitting query queryId=${qContext.queryId} start=${lp.startMs}" +
+        s" end=${lp.endMs} step=${lp.stepMs} splitThresholdMs=$splitThresholdMs splitSizeMs=$splitSizeMs")
+      val numStepsPerSplit = splitSizeMs/lp.stepMs
+      var startTime = lp.startMs
+      var endTime = Math.min(lp.startMs + numStepsPerSplit * lp.stepMs, lp.endMs)
+      val splitPlans: Buffer[ExecPlan] = Buffer.empty
+      while (endTime < lp.endMs ) {
+        splitPlans += planner(copyWithUpdatedTimeRange(lp, TimeRange(startTime, endTime)))
+        startTime = endTime + lp.stepMs
+        endTime = Math.min(startTime + numStepsPerSplit*lp.stepMs, lp.endMs)
+      }
+      // when endTime == lp.endMs - exit condition
+      splitPlans += planner(copyWithUpdatedTimeRange(lp, TimeRange(startTime, endTime)))
+      logger.info(s"splitsize queryId=${qContext.queryId} numWindows=${windowSplits.length}")
+      StitchRvsExec(qContext, stitchDispatcher, splitPlans.toList, parallelChildTasks = false) // Sequential execution
```

Comment:
@sherali42 We may not want to execute all the child plans in parallel. There will be a thundering herd issue here. Instead we need to send subqueries, wait for response before sending more sub queries in.

Reply:
@vishramachandran logic is moved inside SingleClusterPlanner